Generalize `distributeSurplus`.

jonathanknowles opened the pull request:
## Issue Number

ADP-1514

## Summary

This PR adjusts `distributeSurplus` so that:

- It accepts a _list_ of change outputs rather than just a single _optional_ output.
- It accepts change outputs of type `TxOut` rather than `Coin`.
- It handles the responsibility of applying the `fee` and `change` deltas to the original values instead of outsourcing this responsibility to `balanceTransaction`.

We can say that this is a **_generalization_** of `distributeSurplus`, because various implementation details (for example, which change outputs are actually adjusted) are now hidden from callers behind a more general interface, for which we can state more general properties.

This also means that we can write properties that test the **_combination_** of:
- **_computing_** the fee and change deltas
- **_applying_** the fee and change deltas

Whereas before, we only had explicit tests for the former (_computing_ the delta), we can now **_also_** test that:

- The _number_ of change outputs is always preserved.
- The `Coin` `fee` value either stays the same or increases.
- The `Coin` value within each `TxOut` either stays the same or increases.
- The non-ada assets within each `TxOut` are not changed in any way.
- That the deltas are applied correctly.

jonathanknowles committed:
Use `Just` in comment examples for `distributeSurplus`.

The change value is of type `Maybe Coin`.

Therefore, we can also use `Just` in our comment examples.

jonathanknowles committed:
Extract out function `prop_extraFee_coversIncreaseToFeeRequirement`.

While smaller inline properties are perfectly fine, if the definition of
the property becomes too large, then it can be come a challenge to fit
it within the horizontal line limit, resulting in code that is squished
to the right, or code that has excessive wrapping.

This commit moves the definition of an inline property to a top-level
function, giving us more horizontal space to work with. This means that
some statements no longer need to be wrapped. We can unwrap them, which
saves us vertical space.

jonathanknowles committed:
Use `where` clause in `prop_extraFee_coversIncreaseToFeeRequirement`.

Terms defined in `where` clauses can often be indented further to the
left than terms defined in `let` clauses. In this case, it means we no
longer need to wrap the definitions of these terms, which means we can
save vertical space, making the function more concise.

We also remove an unneeded `do`, and an unneeded `$`.

jonathanknowles committed:
Add container types `Empty` and `Solo`.

For the next few commits, we will use these container types to provide
alternative containers for change outputs in `TxFeeAndChange` in cases
where we want to indicate that we have:

- no change outputs
- exactly one change output

Later on, when we parameterize `TxFeeAndChange` further, we will remove
these types.

jonathanknowles committed:
Parameterize the container for change in `TxFeeAndChange`.

jonathanknowles committed:
Add function `mapTxFeeAndChange`.

This function will allow us to transform values of `TxFeeAndChange`
without having to wrap and unwrap record constructors.

jonathanknowles committed:
Make `burnSurplusAsFees` use `Empty` change container type.

Before this commit, this function was only ever called with a change
value of `Nothing`, and only ever returned a change value of `Nothing`.

Therefore, we can safely replace the use of `Maybe` with `Empty`, which
only has a single `Empty` constructor. This clarifies that this function
cannot manipulate change values in any way.

jonathanknowles committed:
Preserve the presence/absence of a change value in `distributeSurplus`.

This means that for a given change value c and output change delta Δc, the
following property will now hold:

>>> isJust c == isJust Δc

This transformation is safe because `balanceTransaction` treats the
following change deltas identically:

- Nothing
- Just (Coin 0)

jonathanknowles committed:
Extract out function `distributeSurplusDeltaWithOneChangeCoin`.

We use the term `Delta`, because this function produces a result that is
a delta: values that must be added on to the original fee and change
outputs in order to produce the final fee and change outputs.

jonathanknowles committed:
Parameterize the type of change in `TxFeeAndChange`.

jonathanknowles committed:
Add functions `distributeSurplusNew` and `distributeSurplusDeltaNew`.

Function `distributeSurplusNew` encapsulates the responsibility of
updating the fee and change values by adding the result of
`distributeSurplusDeltaNew` to the original fee and change values.

We will eventually use `distributeSurplusNew` to replace
`distributeSurplus`.

jonathanknowles committed:
Add `distributeSurplusNew` to `TransactionLayer`.

jonathanknowles committed:
Use `distributeSurplusNew` to simplify `balanceTransaction`.

Since `distributeSurplusNew` includes the logic to update the fee and
change values, we no longer need to perfom this work in
`balanceTransaction`.

jonathanknowles committed:
Use `distributeSurplusDeltaNew` instead of `distributeSurplus` in tests.

In `TransactionSpec`.

jonathanknowles committed:
Remove legacy `distributeSurplus` function.

jonathanknowles committed:
Add property tests for `distributeSurplus`.

These tests cover the case where `distributeSurplus` is successful.

We also add a helper function `txOutSubtractCoin`, in the same style as
`txOutAddCoin`.

jonathanknowles committed:
Add comments to `distributeSurplus` functions.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2310,82 +2339,273 @@ balanceTransactionSpec = do
+  where
+    txOutSubtractCoin :: TxOut -> Coin -> TxOut
+    txOutSubtractCoin txOut toSubtract =
+        over (#tokens . #coin) (`Coin.difference` toSubtract) txOut
```
Perhaps move this to `Types.Tx`, so that it is together with `txOutAddCoin`.

jonathanknowles replied:
Fixed in cda573320d262bd8512e03a45647e9f922ed7711.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -255,34 +257,28 @@ data TransactionLayer k tx = TransactionLayer
         -- Returns `Nothing` for ByronEra transactions.
 
-    , distributeSurplus
+    , minimizeFee
```
The name of this function (and its subordinate functions) should be changed back to `distributeSurplus`.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Util.hs:
```diff
@@ -167,3 +173,47 @@ parseURI (T.unpack -> uri) = runIdentity $ runExceptT $ do
+    Empty <*> Empty = Empty
+
+instance Monad Empty where
+    (>>=) = (const . const) Empty
```
@Anviking might have a way to simplify these definitions.

In any case, we can easily test them with `quickcheck-classes`, as we do for other instances of standard classes.

jonathanknowles replied:
Alternatively, we can make these definitions go away if we make the following transformation:
- `Solo Coin` -> `Coin`
- `Empty Void` -> `()`

Anviking replied:
No code is the best code!

jonathanknowles replied:
Fixed in 374b264546acc7755dacd3f8a1556fda2ef7687d.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1491,33 +1495,64 @@ sizeOfCoin (Coin c)
+        (fee <>)
+        (zipWith (flip txOutAddCoin) change)
+
+minimizeFeeDelta
```
```suggestion
distributeSurplusDelta
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1491,33 +1495,64 @@ sizeOfCoin (Coin c)
-_distributeSurplus feePolicy surplus fc@(TxFeeAndChange fee0 (Just change0)) =
+    -> TxFeeAndChange (Solo Coin)
+    -> Either ErrMoreSurplusNeeded (TxFeeAndChange (Solo Coin))
+minimizeFeeDeltaWithSoloChangeCoin
```
```suggestion
distributeSurplusWithSoloChangeCoin
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1567,13 +1602,13 @@ _distributeSurplus feePolicy surplus fc@(TxFeeAndChange fee0 (Just change0)) =
-    -> TxFeeAndChange
-    -> Either ErrMoreSurplusNeeded TxFeeAndChange
+    -> TxFeeAndChange (Empty Void)
+    -> Either ErrMoreSurplusNeeded (TxFeeAndChange (Empty Void))
```
⚫

jonathanknowles replied:
Fixed in 374b264546acc7755dacd3f8a1556fda2ef7687d.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2299,7 +2304,31 @@ balanceTransactionSpec = do
                     res >= Coin 0 && res <= Coin 8
 
-    describe "distributeSurplus" $ do
+    describe "minimizeFee" $ do
```
```suggestion
    describe "distributeSurplus" $ do
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1491,33 +1495,64 @@ sizeOfCoin (Coin c)
-_distributeSurplus feePolicy surplus fc@(TxFeeAndChange _fee0 Nothing) =
-    burnSurplusAsFees feePolicy surplus fc
-_distributeSurplus feePolicy surplus fc@(TxFeeAndChange fee0 (Just change0)) =
+    -> TxFeeAndChange (Solo Coin)
```
Thinking about it, perhaps this could be just `Coin`:

```suggestion
    -> TxFeeAndChange Coin
```

jonathanknowles replied:
Fixed in 374b264546acc7755dacd3f8a1556fda2ef7687d.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1567,13 +1602,13 @@ _distributeSurplus feePolicy surplus fc@(TxFeeAndChange fee0 (Just change0)) =
     -> Coin -- Surplus
-    -> TxFeeAndChange
-    -> Either ErrMoreSurplusNeeded TxFeeAndChange
+    -> TxFeeAndChange (Empty Void)
```
We could also replace `Empty Void` with `()`.

jonathanknowles replied:
Fixed in 374b264546acc7755dacd3f8a1556fda2ef7687d.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1491,33 +1495,64 @@ sizeOfCoin (Coin c)
+            burnSurplusAsFees feePolicy surplus
+                (TxFeeAndChange fee Empty)
+            <&> txFeeAndChange id
+                (\Empty -> [])
```
You can still use an explicit pattern match after dropping `Empty`:
```suggestion
                (\() -> [])
```

jonathanknowles replied:
Fixed in 374b264546acc7755dacd3f8a1556fda2ef7687d.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2310,82 +2339,277 @@ balanceTransactionSpec = do
+            ]
+        slope <- frequency
+            [ (1, pure 0.0)
+            , (3, getPositive <$> arbitrary)
```
💯 thanks!

It might be nice to throw in a few mainnet values:
```suggestion
            , (3, getPositive <$> arbitrary)
            , (1, pure 44)
```

And also for the intercept:
```
            , (1, pure 150_000)
```

jonathanknowles replied:
Nice idea!

Fixed in bd4b985ec2000776776df7d08812830cacc90918.

Anviking commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -477,3 +505,31 @@ data ErrUpdateSealedTx
+
+-- | Manipulates a 'TxFeeAndChange' value.
+--
+txFeeAndChange
```
I think it's worth calling this `mapTxFeeAndChange`. While `either` and `maybe` are well known, I don't think we use similar naming ourselves. E.g. `mapChainFollower`, `mapChainFollowerLog`

For brevity we could also have a separate `mapChange`, which would cover most of the use-cases in the implementation.

jonathanknowles replied:
Fair enough! Fixed in 1ba4983911176d6bbd78269ad0eae9dca60da6c9.

(I adjusted the original commit that added this function rather than making a new commit).

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1477,6 +1495,123 @@ sizeOfCoin (Coin c)
+                (TxFeeAndChange fee (Solo firstChange))
+            <&> txFeeAndChange id
+                ((: replicate (length change - 1) (Coin 0)) . unSolo)
+        Nothing ->
```
Could also do this:
```suggestion
    case change of
         firstChange : rest ->
            distributeSurplusDeltaWithSoloChangeCoin feePolicy surplus
                (TxFeeAndChange fee (Solo firstChange))
            <&> txFeeAndChange id
                (: map (const $ Coin 0) rest)
        [] ->
```

jonathanknowles replied:
Nice idea! Fixed in d5ae260d5045ac2f859ba43d0394476ae032ef05.

jonathanknowles committed:
Remove `New` suffix from new `distributeSurplus` function family.

The legacy function has been removed, so we can now perform the
following renamings:

- `distributeSurplusNew`      -> `distributeSurplus`
- `distributeSurplusDeltaNew` -> `distributeSurplusDelta`

jonathanknowles committed:
Revise `prop_distributeSurplusDelta_coversCostIncreaseAndConservesSurplus`.

We add a longer comment to explain the purpose of the property.

jonathanknowles committed:
Remove usages of `Empty` and `Solo`.

These types were introduced because we parameterized the **container type**
for change in `TxFeeAndChange`:

```hs
data TxFeeAndChange f = TxFeeAndChange
    { fee :: Coin
    , change :: f Coin
    }
```

This required us to instantiate `f` with types of kind `* -> *`.

But later on we parameterized the type of `change` so that it can be any type
of kind `*`:

```hs
data TxFeeAndChange change = TxFeeAndChange
    { fee :: Coin
    , change :: change
    }
```

Hence we can make the following simplifications:

  - Empty Void -> ()
  - Solo  Coin -> Coin

And therefore we can remove the `Empty` and `Solo` types.

In response to review discussion:

https://github.com/input-output-hk/cardano-wallet/pull/3238#discussion_r853105496

jonathanknowles committed:
Simplify `distributeSurplusDelta` by pattern matching directly on change list.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3238#discussion_r853497227

jonathanknowles committed:
Also test with mainnet fee policy in `prop_distributeSurplus_onSuccess`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3238#discussion_r853330566

Anviking reviewed (approved):
lgtm, thanks!

Anviking committed:
Use Arbitrary feePolicy in distributeSurplusDelta property
